pennylane_lightning/core/src/simulators/lightning_kokkos/measurements/MeasurementsKokkos.hpp

```diff
@@ -477,26 +476,31 @@ class Measurements final
     /**
      * @brief Probabilities to measure rotated basis states.
      *
-     * @return Floating point std::vector with probabilities
+     * @return Kokkos::View with probabilities
      * in lexicographic order.
      */
-    auto probs() -> std::vector<PrecisionT> {
+    auto probs_core() -> Kokkos::View<PrecisionT *> {
         const std::size_t N = this->_statevector.getLength();
-
-        Kokkos::View<ComplexT *> arr_data = this->_statevector.getView();
-        Kokkos::View<PrecisionT *> d_probability("d_probability", N);
-
-        // Compute probability distribution from StateVector
+        auto sv = this->_statevector.getView();
+        Kokkos::View<PrecisionT *> d_probs("d_probs", N);
         Kokkos::parallel_for(
             Kokkos::RangePolicy<KokkosExecSpace>(0, N),
-            getProbFunctor<PrecisionT>(arr_data, d_probability));
-
-        std::vector<PrecisionT> probabilities(N, 0);
+            KOKKOS_LAMBDA(const std::size_t k) {
+                const PrecisionT rsv = sv(k).real();
+                const PrecisionT isv = sv(k).imag();
+                d_probs(k) = rsv * rsv + isv * isv;
+            });
+        return d_probs;
+    }
 
-        Kokkos::deep_copy(UnmanagedPrecisionHostView(probabilities.data(),
-                                                     probabilities.size()),
-                          d_probability);
-        return probabilities;
+    /**
+     * @brief Probabilities to measure rotated basis states.
+     *
+     * @return Floating point std::vector with probabilities
+     * in lexicographic order.
+     */
+    auto probs() -> std::vector<PrecisionT> {
+        return view2vector(probs_core());
```

Comment:
Can we pass the data directly to python layer given the computation is with cpus?

Reply:
`probs_core` is performance on the Kokkos parallel space (CPU or GPU).

Reply:
Good point, we were doing that everywhere but it is indeed wasteful. Now that the operations are nicely packaged in `view2vector` we could try to skip copies and roll out the change everywhere.

Reply:
On second thought, I think we make the deep copy because a view is like a pointer, and hence we want to make sure the underlying data won't be changed by something else. We could make shallow copies in certain cases, but let's keep it safe for now in the interest of time.